File: C++/Search in Rotated Sorted Array.cpp
class Solution {
private:
    int dfs(vector<int>& nums, int target, int left, int right){
        if(left > right)
            return -1;
        
        int mid = (int) ((left + right) / 2);
        
        if(nums[mid] == target)
            return mid;
        else if(nums[left] <= target and target < nums[mid])
            return dfs(nums, target, left, mid - 1);
        else if(nums[mid] < target and target <= nums[right])
            return dfs(nums, target, mid + 1, right);
        else if(nums[left] > nums[mid])
            return dfs(nums, target, left, mid - 1);
        else
            return dfs(nums, target, mid + 1, right);
    }
public:
    int search(vector<int>& nums, int target) {
        return dfs(nums, target, 0, nums.size() - 1);
    }
};

File: Python/Search in Rotated Sorted Array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        
        if not nums:
            return -1
        
        def binarySearch(nums, target, left, right):
            if left > right:
                return -1
            
            mid = int((left + right) / 2)
            if nums[mid] == target:
                return mid
            elif nums[left] <= nums[right]:
                if nums[mid] < target:
                    return binarySearch(nums, target, mid + 1, right)
                else:
                    return binarySearch(nums, target, left, mid - 1)
            else:
                if nums[mid] < target:
                    if target <= nums[right]:
                        return binarySearch(nums, target, mid + 1, right)
                    else:
                        if nums[mid] >= nums[left]:
                            return binarySearch(nums, target, mid + 1, right)
                        else:
                            return binarySearch(nums, target, left, mid - 1)
                else:
                    if nums[mid] <= nums[right]:
                        return binarySearch(nums, target, left, mid - 1)
                    else:
                        if target <= nums[right]:
                            return binarySearch(nums, target, mid + 1, right)
                        else:
                            return binarySearch(nums, target, left, mid - 1)
        
        return binarySearch(nums, target, 0, len(nums) - 1)